sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -496,8 +496,8 @@ trait Column extends DataFrame {
    * @param startPos expression for the starting position.
    * @param len expression for the length of the substring.
    */
-  def substr(startPos: Column, len: Column): Column = {
-    new IncomputableColumn(Substring(expr, startPos.expr, len.expr))
+  def substr(startPos: Column, len: Column): Column = constructColumn(null) {
```

Comment:
this change breaks this pr : https://github.com/apache/spark/pull/4332

You need to use 
```
exprToColumn(Substring(expr, startPos.expr, len.expr), computable = false)
```